### ISSUE TITLE
OnDisconnectAsync() does not get called after timeout if client loses connection in the first 20 seconds of the connection.

### ISSUE THREAD

**Issue by alagoa** [aspnet/AspNetCore#8153]
## [`Code here`](https://github.com/alagoa/SignalRTest)

Using SignalR 2.4.0 [EDIT: mistake, it's SignalR Core 1.1.0]. I have a simple Hub that only has 2 methods, `OnConnectedAsync()` and `OnDisconnectedAsync()`.

```
namespace SignalRTest.Hubs
{
    public class MyHub : Hub
    {
        public override Task OnConnectedAsync()
        {
            Globals.SystemLogger.LogTrace($"Starting onConnectedAsync for user with connectionId {Context.ConnectionId}.");
            return base.OnConnectedAsync();
        }

        public override Task OnDisconnectedAsync(Exception e)
        {
            Globals.SystemLogger.LogTrace($"Starting onDisconnectedAsync for user with connectionId {Context.ConnectionId}.");
            return base.OnDisconnectedAsync(e);
        }
    }
}
```

## Describe the bug

When using use a mobile application (tested on Android 8 and 9) to connect to the Hub, if one disconnects (by switching off WiFi, **not** by stopping the application),  within the first ~20 seconds after the connections was established, the server will not call the `OnDisconnectAsync()` method after the timeout (default is 30 seconds) and will behave as if the client was still connected. Instead, `OnDisconnectAsync()` only gets called **when the client turns the WiFi/data back on**, no matter how long it takes.

## Steps to reproduce

1. Launch the server
2. Connect to the hub (port is 53353, hub name is `general`)
3. Switch off WiFi/data within the first 20 seconds of the connection.

The `onDisconnectAsync()` won't get called after the timeout passes. The server will still try to ping the client during the timeout window, but after that, no more packets will be sent.

### More information

If the same use case occurs, but the client disconnects **after** the first 20 seconds, the server will call the `OnDisconnectAsync()` and close the connection on their side, as expected. Here is the log for that case (the client disconnects at around 18:00:05):

**Log 1 (expected behavior)**
```
[2019/03/04 17:59:43.458] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0FUV5NBJF" started.
[2019/03/04 17:59:43.517] [INFO] [Microsoft.AspNetCore.Hosting.Internal.WebHost] Request starting HTTP/1.1 POST http://192.168.51.147:53353/general/negotiate?uid=de4d639c-547d-4118-939b-65ce3c37b2d1  0
[2019/03/04 17:59:43.518] [DBUG] [Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware] Wildcard detected, all requests with hosts will be allowed.
[2019/03/04 17:59:43.520] [WARN] [Microsoft.AspNetCore.HttpsPolicy.HttpsRedirectionMiddleware] Failed to determine the https port for redirect.
[2019/03/04 17:59:43.527] [DBUG] [Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware] POST requests are not supported
[2019/03/04 17:59:43.550] [DBUG] [Microsoft.AspNetCore.Routing.Matching.DfaMatcher] No candidates found for the request path '/general/negotiate'
[2019/03/04 17:59:43.550] [DBUG] [Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware] Request did not match any endpoints
[2019/03/04 17:59:43.560] [DBUG] [Microsoft.AspNetCore.Routing.RouteBase] Request successfully matched the route with name '(null)' and template '/general/negotiate'
[2019/03/04 17:59:43.568] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionManager] New connection omx31k-ETJvlhsaAtIScKQ created.
[2019/03/04 17:59:43.594] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher] Sending negotiation response.
[2019/03/04 17:59:43.596] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0FUV5NBJF" completed keep alive response.
[2019/03/04 17:59:43.599] [INFO] [Microsoft.AspNetCore.Hosting.Internal.WebHost] Request finished in 82.889ms 200 application/json
[2019/03/04 17:59:43.719] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0FUV5NBJG" started.
[2019/03/04 17:59:43.738] [INFO] [Microsoft.AspNetCore.Hosting.Internal.WebHost] Request starting HTTP/1.1 GET http://192.168.51.147/general?uid=de4d639c-547d-4118-939b-65ce3c37b2d1&id=omx31k-ETJvlhsaAtIScKQ  0
[2019/03/04 17:59:43.738] [DBUG] [Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware] The request path /general does not match a supported file type
[2019/03/04 17:59:43.740] [DBUG] [Microsoft.AspNetCore.Routing.Matching.DfaMatcher] No candidates found for the request path '/general'
[2019/03/04 17:59:43.740] [DBUG] [Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware] Request did not match any endpoints
[2019/03/04 17:59:43.741] [DBUG] [Microsoft.AspNetCore.Routing.RouteBase] Request successfully matched the route with name '(null)' and template '/general'
[2019/03/04 17:59:43.755] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher] Establishing new connection.
[2019/03/04 17:59:43.761] [DBUG] [Microsoft.AspNetCore.SignalR.HubConnectionHandler] OnConnectedAsync started.
[2019/03/04 17:59:43.766] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport] Socket opened using Sub-Protocol: '(null)'.
[2019/03/04 17:59:43.834] [DBUG] [Microsoft.AspNetCore.SignalR.Internal.DefaultHubProtocolResolver] Found protocol implementation for requested protocol: json.
[2019/03/04 17:59:43.836] [INFO] [Microsoft.AspNetCore.SignalR.HubConnectionContext] Completed connection handshake. Using HubProtocol 'json'.
[2019/03/04 17:59:43.843] [TRCE] [sys] Starting onConnectedAsync for user with connectionId omx31k-ETJvlhsaAtIScKQ.
[2019/03/04 18:00:03.681] [DBUG] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] Connection id "0HLL0FUV5NBJF" received FIN.
[2019/03/04 18:00:03.683] [DBUG] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] Connection id "0HLL0FUV5NBJF" sending FIN because: "The client closed the connection."
[2019/03/04 18:00:03.692] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0FUV5NBJF" disconnecting.
[2019/03/04 18:00:03.695] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0FUV5NBJF" stopped.
[2019/03/04 18:00:45.388] [DBUG] [Microsoft.AspNetCore.SignalR.HubConnectionContext] Client timeout (30000ms) elapsed without receiving a message from the client. Closing connection.
[2019/03/04 18:00:45.415] [TRCE] [sys] Starting onDisconnectedAsync for user with connectionId omx31k-ETJvlhsaAtIScKQ.
[2019/03/04 18:00:45.415] [DBUG] [Microsoft.AspNetCore.SignalR.HubConnectionHandler] OnConnectedAsync ending.
[2019/03/04 18:00:45.442] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport] Waiting for the client to close the socket.
[2019/03/04 18:00:46.384] [FAIL] [Microsoft.AspNetCore.SignalR.HubConnectionContext] Failed writing message. Aborting connection.
[2019/03/04 18:00:50.448] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport] Socket closed.
[2019/03/04 18:00:50.467] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionManager] Removing connection omx31k-ETJvlhsaAtIScKQ from the list of connections.
[2019/03/04 18:00:50.481] [INFO] [Microsoft.AspNetCore.Hosting.Internal.WebHost] Request finished in 66743.2074ms 101 
[2019/03/04 18:00:50.488] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0FUV5NBJG" disconnecting.
[2019/03/04 18:00:50.488] [DBUG] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] Connection id "0HLL0FUV5NBJG" sending FIN because: "The Socket transport's send loop completed gracefully."
[2019/03/04 18:00:50.501] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0FUV5NBJG" stopped.
```

And here is the packet capture for that case (client disconnects shortly after packet # 230):

**Server**: 192.168.51.147
**Client**: 192.168.51.145

**Fig. 1 (expected behavior)**
![image](https://user-images.githubusercontent.com/12493763/53726726-ec884d80-3eb1-11e9-9f53-d4181174a932.png) 

As in can be seen in **Fig. 1**, the server sends a Websocket Connection Close packet. This happens at roughly the same time as the lines with the timestamp *18:00:45* in **Log 1**.

The 20 seconds mark is also when packets # 228 - # 230 are exchanged. They correspond to the lines with the timestamp *18:00:03* in **Log 1**

__________

However, for the use case presented, where the client disconnects before those packets are exchanged, the connection will never be closed by the server. Here is the log:

 **Log 2**
```
[2019/03/04 18:03:40.057] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0G15M2DAO" started.
[2019/03/04 18:03:40.116] [INFO] [Microsoft.AspNetCore.Hosting.Internal.WebHost] Request starting HTTP/1.1 POST http://192.168.51.147:53353/general/negotiate?uid=de4d639c-547d-4118-939b-65ce3c37b2d1  0
[2019/03/04 18:03:40.118] [DBUG] [Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware] Wildcard detected, all requests with hosts will be allowed.
[2019/03/04 18:03:40.121] [WARN] [Microsoft.AspNetCore.HttpsPolicy.HttpsRedirectionMiddleware] Failed to determine the https port for redirect.
[2019/03/04 18:03:40.128] [DBUG] [Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware] POST requests are not supported
[2019/03/04 18:03:40.149] [DBUG] [Microsoft.AspNetCore.Routing.Matching.DfaMatcher] No candidates found for the request path '/general/negotiate'
[2019/03/04 18:03:40.150] [DBUG] [Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware] Request did not match any endpoints
[2019/03/04 18:03:40.158] [DBUG] [Microsoft.AspNetCore.Routing.RouteBase] Request successfully matched the route with name '(null)' and template '/general/negotiate'
[2019/03/04 18:03:40.162] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionManager] New connection lQGffy0ZQCuRRwaPbW-Obg created.
[2019/03/04 18:03:40.188] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher] Sending negotiation response.
[2019/03/04 18:03:40.191] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0G15M2DAO" completed keep alive response.
[2019/03/04 18:03:40.194] [INFO] [Microsoft.AspNetCore.Hosting.Internal.WebHost] Request finished in 77.4413ms 200 application/json
[2019/03/04 18:03:40.322] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0G15M2DAP" started.
[2019/03/04 18:03:40.341] [INFO] [Microsoft.AspNetCore.Hosting.Internal.WebHost] Request starting HTTP/1.1 GET http://192.168.51.147/general?uid=de4d639c-547d-4118-939b-65ce3c37b2d1&id=lQGffy0ZQCuRRwaPbW-Obg  0
[2019/03/04 18:03:40.341] [DBUG] [Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware] The request path /general does not match a supported file type
[2019/03/04 18:03:40.344] [DBUG] [Microsoft.AspNetCore.Routing.Matching.DfaMatcher] No candidates found for the request path '/general'
[2019/03/04 18:03:40.344] [DBUG] [Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware] Request did not match any endpoints
[2019/03/04 18:03:40.345] [DBUG] [Microsoft.AspNetCore.Routing.RouteBase] Request successfully matched the route with name '(null)' and template '/general'
[2019/03/04 18:03:40.383] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher] Establishing new connection.
[2019/03/04 18:03:40.413] [DBUG] [Microsoft.AspNetCore.SignalR.HubConnectionHandler] OnConnectedAsync started.
[2019/03/04 18:03:40.424] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport] Socket opened using Sub-Protocol: '(null)'.
[2019/03/04 18:03:40.500] [DBUG] [Microsoft.AspNetCore.SignalR.Internal.DefaultHubProtocolResolver] Found protocol implementation for requested protocol: json.
[2019/03/04 18:03:40.503] [INFO] [Microsoft.AspNetCore.SignalR.HubConnectionContext] Completed connection handshake. Using HubProtocol 'json'.
[2019/03/04 18:03:40.509] [TRCE] [sys] Starting onConnectedAsync for user with connectionId lQGffy0ZQCuRRwaPbW-Obg.
[2019/03/04 18:05:40.453] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0G15M2DAO" disconnecting.
[2019/03/04 18:05:40.459] [DBUG] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] Connection id "0HLL0G15M2DAO" sending FIN because: "The Socket transport's send loop completed gracefully."
[2019/03/04 18:05:40.477] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0G15M2DAO" stopped.
[2019/03/04 18:05:42.795] [DBUG] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] Connection id "0HLL0G15M2DAP" reset.
[2019/03/04 18:05:42.795] [DBUG] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] Connection id "0HLL0G15M2DAP" sending FIN because: "The client closed the connection."
[2019/03/04 18:05:42.795] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0G15M2DAP" disconnecting.
[2019/03/04 18:05:42.798] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport] Socket connection closed prematurely.
[2019/03/04 18:05:42.800] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport] Waiting for the application to finish sending data.
[2019/03/04 18:05:42.801] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport] Socket closed.
[2019/03/04 18:05:42.812] [TRCE] [sys] Starting onDisconnectedAsync for user with connectionId lQGffy0ZQCuRRwaPbW-Obg.
[2019/03/04 18:05:42.812] [DBUG] [Microsoft.AspNetCore.SignalR.HubConnectionHandler] OnConnectedAsync ending.
[2019/03/04 18:05:42.818] [DBUG] [Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionManager] Removing connection lQGffy0ZQCuRRwaPbW-Obg from the list of connections.
[2019/03/04 18:05:42.823] [INFO] [Microsoft.AspNetCore.Hosting.Internal.WebHost] Request finished in 122482.1717ms 101 
[2019/03/04 18:05:42.823] [DBUG] [Microsoft.AspNetCore.Server.Kestrel] Connection id "0HLL0G15M2DAP" stopped.
```

And the packet capture (client disconnects shortly after packet # 264):

 **Fig. 2**
![image](https://user-images.githubusercontent.com/12493763/53727335-4ccbbf00-3eb3-11e9-9116-c351343ffe9b.png)

In the problematic case, the two [FIN, ACK] packets present in **Fig. 1** will not get exchanged, since the client does not have access to the network. As said above, the server tries to contact the client, but never closes the socket like in **Fig. 1**. When the client reconnects (about 50 seconds after packet # 509), it sends a **TCP Reset** to the server, to tell it to close the websocket. The server closes it and calls OnDisconnectAsync().

So, my question is, should SignalR behave this way? If so, why? Why are those [FIN, ACK] packets important and why isn't the timeout respected? In the docs it's stated that `It could take longer than this timeout interval for the client to actually be marked disconnected, due to how this is implemented.`, but from my tests, the client won't be marked as disconnected at all in this particular case.

Thanks in advance.

**Comment by halter73** [comment]
What version of the server and client are you using. It's interesting there are no logs related to missed keep alives.

**Comment by alagoa** [comment]
Thanks for the quick reply.
I was using SignalR Core `2.4.0` on the server and the client was using version `1.0.4`. After updating the client version to `1.1.0` the issue was fixed.
Can anyone give me more insight into what was exactly changed between the two versions?

**Comment by BrennanConroy** [comment]
First off 2.4.0 doesn't exist. Secondly, in 1.1.0 we added keep alives from the client to the server. So now if the server detects no keep alive then it will disconnect the client.

**Comment by alagoa** [comment]
My bad, I'm using SignalR Core 1.1.0 on the server.

I get why those changes make the problem disappear, but I still don't understand how that relates to this particular issue, though.
I don't understand why the server behaves differently in those two situations I presented. In both cases, the server tries to reach the client, but it only closes the connection in one of them, why?

**Comment by davidfowl** [comment]
In 1.1.0 of the client, we added client to server pings to aggressively detect disconnections on the server.

**Comment by TanukiSharp** [comment]
Thanks for the info @davidfowl (and congrats for your recent promotion !)

I've read somewhere ASP.NET Core 3 will allow to use a different JSON library than NewtonSoft. Any chance this happens too with SignalR soon ? We are using it (SignalR client) in Unity3D but because of iOS and NewtonSoft JSON clash we had to fallback using BestHTTP plugin, which seems to still be on 1.0.4.

**Comment by davidfowl** [comment]
> Thanks for the info @davidfowl (and congrats for your recent promotion !)

Thanks!

> I've read somewhere ASP.NET Core 3 will allow to use a different JSON library than NewtonSoft. Any chance this happens too with SignalR soon ? We are using it (SignalR client) in Unity3D but because of iOS and NewtonSoft JSON clash we had to fallback using BestHTTP plugin, which seems to still be on 1.0.4.

Yes, we're doing work on this in 3.0, it probably won't solve your unity problems on iOS though. At least not in the initial release.

**Comment by halter73** [comment]
Closing since this should be addressed by client pings in 1.1.
